teps/0075-object-param-and-result-types.md

```diff
@@ -0,0 +1,431 @@
+- [Upgrade &amp; Migration Strategy](#upgrade--migration-strategy)
+- [Implementation Pull request(s)](#implementation-pull-request-s)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+_Recommendation: read
+[TEP-0076 (array support in results and indexing syntax)](https://github.com/tektoncd/community/pull/477)
+before this TEP, as this TEP builds on that one._
+
+This TEP proposes adding dictionary (aka object) types to Tekton Task and Pipeline
+[results](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#emitting-results)
+and [params](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#specifying-parameters),
+as well as adopting a small subset of [JSONPath](https://tools.ietf.org/id/draft-goessner-dispatch-jsonpath-00.html)
+syntax (precedent in
+[the array expansion syntax we are using](jhttps://github.com/tektoncd/pipeline/issues/1393#issuecomment-561476075)) 
+for accessing values in these dictionaries.
+
+This proposal is dependent on
+[TEP-0076 (array support in results and indexing syntax)](https://github.com/tektoncd/community/pull/477)
+in that this TEP should follow the precedents set there and also in that if we decide not to support array results, we
+probably won't support dictionary results either.
+
+This proposal also includes adding support for limited use of
+[json object schema](https://json-schema.org/understanding-json-schema/reference/object.html),
+to express the expected structure of the dictionary.
+
+_Dictionary vs. object: The intended feature was supporting "dictionaries", but json schema calls these "objects" so
+this proposal tries to use "objects" as the term for this type._
+
+## Motivation
+
+Tasks declare workspaces, params, and results, and these can be linked in a Pipeline, but external tools looking at
+these Pipelines cannot reliably tell when images are being built, or git repos are being used. Current ways around this
+are:
+* [PipelineResources](https://github.com/tektoncd/pipeline/blob/main/docs/resources.md) (proposed to be
+  deprecated in TEP-0074)
+* Defining param names with special meaning, for example
+  [Tekton Chains type hinting](https://github.com/tektoncd/chains/blob/main/docs/config.md#chains-type-hinting))
+
+This proposal takes the param name type hinting a step further by introducing object types for results: this will
+allow Tasks to define structured results. These structured results can be used to define interfaces, for example the
+values that are expected to be produced by Tasks which do `git clone`. This is possible with just string results, but
+without a grouping mechanisms, all results and params are declared at the same level and it is not obvious which results
+are complying with a general interface and which are specific to the task.
+
+### Goals
+
+* Tidy up Task interfaces by allowing Tasks to group related parameters (similar to the
+  [long parameter list](https://www.arhohuttunen.com/long-parameter-list/) or
```

Comment:
Agreed, but I think it's probably on the "consumer" side of things that it might reduce. For example, the `git-clone` returns a dictonnary, and I can pass that dictonnary as is (because it "fits") on another task, in that case it reduce the yaml (*but* it means tboth task need to take that dictionnary, which may introduce the same type of dependency between task that we had with `PipelineResource`)

Reply:
Ah innnnteresting, I hadn't thought about that, good point!